### PR TITLE
tree2: Split FieldKind class, shrink package API

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Split Editor logic from FieldKind into package private subclass, and down cast inside the package as needed to access it.

This removes a lot of editing related types from the package API, including the coded types causing issues in https://github.com/microsoft/FluidFramework/pull/17231.

## Breaking Changes

Many types were removed from the API: these should all be implementation details of specific field kinds which should not have been accessed outside the package.
 
## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/modular-schema/fieldKind.ts [hunk 1]
```diff
@@ -28,12 +28,45 @@ import { FieldChangeHandler, FieldEditor } from "./fieldChangeHandler";
  * These policies include the data encoding, change encoding, change rebase and change application.
  *
  * @sealed @alpha
+ * @privateRemarks
+ * This being @sealed is for users of this package.
+ * This package itself may provide implementations.
+ * This pattern was picked instead of an interface since we do not have a convention for how to mark interfaces as only allowed to be implemented by the package declaring them,
+ * and using a class also gets stronger. (nominal) typing for objects which will be down cast.
```
Comment (Abe27342) [inline experimental/dds/tree2/src/feature-libraries/modular-schema/fieldKind.ts:35]:
```suggestion
 * and using a class also gets stronger (nominal) typing for objects which will be down cast.
```

**Review by Abe27342 — APPROVED** [review]

**Review by Abe27342 — APPROVED** [review]
Looks great!

**Review by Abe27342 — APPROVED** [review]
Looks great!

File: experimental/dds/tree2/src/feature-libraries/modular-schema/fieldKind.ts (hunk 1 above)
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/modular-schema/fieldKind.ts:35]:
```suggestion
 * and using a class also gets stronger (nominal) typing for objects which will be down cast.
```

**Pull request merged**